pandas/core/groupby.py

```diff
@@ -2850,7 +2851,25 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
     elif isinstance(key, BaseGrouper):
         return key, [], obj
 
-    # Everything which is not a list is a key (including tuples):
+    # In the future, a tuple key will always mean an actual key,
+    # not an iterable of keys. In the meantime, we attempt to provide
+    # a warning. We can assume that the user wanted a list of keys when
+    # the key is not in the index. We just have to be careful with
+    # unhashble elements of `key`. Any unhashable elements implies that
+    # they wanted a list of keys.
+    # https://github.com/pandas-dev/pandas/issues/18314
+    is_tuple = isinstance(key, tuple)
+    all_hashable = is_tuple and all(is_hashable(x) for x in key)
+
+    if is_tuple:
+        if not all_hashable or key not in obj:
```

Comment:
I'm lost. Why do you check that elements are _not_ hashable? I would have done instead
```python
if all_hashable and key not in obj and set(key).issubset(obj):
```
or (if we want to account for the to-be-deprecated possibility to index with missing keys):
```python
if all_hashable and key not in obj and set(key) & (obj):
```

Reply:
Or better - performance-wise:
```python
if key not in obj and all(is_hashable(x) for x in key) and set(key).issubset(obj):
```

Reply:
This is for the case where you're grouping by non-hashable arrays like in https://github.com/pandas-dev/pandas/issues/18314#issue-274315173

In that case, don't we know that they're certainly relying on `groupby((a, b))` to be `groupby([a, b])`, so we want to warn and listify?

Reply:
I do still need to handle your KeyError example.

Reply:
Ahh, I see what you're doing now. Yes, that's probably better, and will make handling the KeyError easier.